### PR TITLE
Migeng 211 multi json tar

### PR DESCRIPTION
**Pull request by jonathanvila**
https://jira.coreos.com/browse/MIGENG-211

Addin an aggregator to sum the values for ICS

### CODE REVIEW

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 1]
```diff
@@ -162,34 +162,47 @@ public void configure() {
...
-                        .to("direct:calculate")
+                        .split(new TarSplitter()).aggregationStrategy(this::calculateICSAggregated)
+                            .streaming()
+                            .to("direct:calculate")
+                        .end()
                     .endChoice()
                     .otherwise()
                         .to("direct:calculate")
-                .end();
+                        .process(exchange -> exchange.getIn().getHeaders().put("uploadformdata", exchange.getIn().getBody(UploadFormInputDataModel.class)))
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: Please make `uploadformdata` a static field like `ANALYSIS_ID` and reference it where needed

File: src/main/java/org/jboss/xavier/integrations/migrationanalytics/business/ParamsCalculator.java [hunk 2]
```diff
@@ -17,11 +18,14 @@
...
 
+    @Inject
+    private AnalysisService analysisService;
+
     private static Integer calculateHypervisors(Object e, String cpuTotalCoresPath, String cpuCoresPerSocketPath) {
         Map mapa = (Map) e;
         Integer cputotalcores = (Integer) mapa.get(cpuTotalCoresPath);
         Integer cpucorespersocket = (Integer) mapa.get(cpuCoresPerSocketPath);
-        return cputotalcores / (cpucorespersocket * 2);
+        return Double.valueOf(Math.ceil(cputotalcores / (cpucorespersocket * 2.0))).intValue();
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/migrationanalytics/business/ParamsCalculator.java]: The boxing to Double and the `intValue()` could be avoided casting to `int`? 
something like `return (int) Math.ceil(cputotalcores / (cpucorespersocket * 2.0));`should work

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 3]
```diff
@@ -162,34 +162,47 @@ public void configure() {
...
+                .end()
+                // at this time it's safe to generate the WSR
+                .to("direct:send-costsavings")
+                .to("direct:calculate-workloadsummaryreportmodel");
 
         from("direct:calculate")
                 .id("calculate")
                 .convertBodyTo(String.class)
-                .multicast()
+                .multicast().aggregationStrategy(new org.apache.camel.processor.aggregate.GroupedBodyAggregationStrategy())
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: please remove the FQN and use an `import`

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 4]
```diff
@@ -162,34 +162,47 @@ public void configure() {
...
                     .to("direct:calculate-costsavings", "direct:calculate-vmworkloadinventory", "direct:flags-shared-disks")
+                .setBody(e -> e.getIn().getBody(List.class).get(0))
                 .end();
 
-        from("direct:calculate-costsavings")
-                .id("calculate-costsavings")
-                .transform().method("calculator", "calculate(${body}, ${header.MA_metadata})")
+
+        from("direct:send-costsavings").id("send-costsavings")
+                .log("SEND-COSTSAVINGS")
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: Could you add the `LoggingLevel.DEBUG` param please?

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 5]
```diff
@@ -162,34 +162,47 @@ public void configure() {
...
+                .setBody(header("uploadformdata"))
                 .log("Message to send to AMQ : ${body}")
                 .to("jms:queue:uploadFormInputDataModel")
                 .end();
 
+        from("direct:calculate-costsavings").id("calculate-costsavings")
+                .log("COST SAVINGS")
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: Could you add the `LoggingLevel.DEBUG` param please?

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 6]
```diff
@@ -162,34 +162,47 @@ public void configure() {
...
+                .setBody(e -> e.getIn().getBody(List.class).get(0))
                 .end();
 
-        from("direct:calculate-costsavings")
-                .id("calculate-costsavings")
-                .transform().method("calculator", "calculate(${body}, ${header.MA_metadata})")
+
+        from("direct:send-costsavings").id("send-costsavings")
+                .log("SEND-COSTSAVINGS")
+                .setBody(header("uploadformdata"))
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: See above about using static field

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 7]
```diff
@@ -209,6 +222,15 @@ public void configure() {
                 .process(httpError403());
     }
 
+    private Exchange calculateICSAggregated(Exchange old, Exchange neu) {
+            UploadFormInputDataModel newObj = (UploadFormInputDataModel) neu.getIn().getBody(List.class).get(0);
+            UploadFormInputDataModel oldObj = (old != null) ? old.getIn().getHeader("uploadformdata", UploadFormInputDataModel.class) : null;
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: See above about using static field

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 8]
```diff
@@ -209,6 +222,15 @@ public void configure() {
...
+            newObj.setTotalDiskSpace(newObj.getTotalDiskSpace() + ((old != null) ? oldObj.getTotalDiskSpace() : 0));
+            newObj.setHypervisor(newObj.getHypervisor() + ((old != null) ? oldObj.getHypervisor() : 0));
+            neu.getIn().getHeaders().put("uploadformdata", newObj);
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: See above about using static field

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 9]
```diff
@@ -209,6 +222,15 @@ public void configure() {
                 .process(httpError403());
     }
 
+    private Exchange calculateICSAggregated(Exchange old, Exchange neu) {
+            UploadFormInputDataModel newObj = (UploadFormInputDataModel) neu.getIn().getBody(List.class).get(0);
+            UploadFormInputDataModel oldObj = (old != null) ? old.getIn().getHeader("uploadformdata", UploadFormInputDataModel.class) : null;
+            newObj.setTotalDiskSpace(newObj.getTotalDiskSpace() + ((old != null) ? oldObj.getTotalDiskSpace() : 0));
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: Should this check be  `oldObj != null`?
```suggestion
            newObj.setTotalDiskSpace(newObj.getTotalDiskSpace() + ((oldObj != null) ? oldObj.getTotalDiskSpace() : 0));
```

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java [hunk 10]
```diff
@@ -209,6 +222,15 @@ public void configure() {
...
+            newObj.setHypervisor(newObj.getHypervisor() + ((old != null) ? oldObj.getHypervisor() : 0));
```
Comment (mrizzi) [inline src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: See above
```suggestion
            newObj.setHypervisor(newObj.getHypervisor() + ((oldObj != null) ? oldObj.getHypervisor() : 0));
```

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -66,7 +66,7 @@ camel.component.servlet.mapping.context-path=/api/xavier/*
 # WIR parallel creation
 parallel.wir=true
 
-insights.properties=yearOverYearGrowthRatePercentage,percentageOfHypervisorsMigratedOnYear1,percentageOfHypervisorsMigratedOnYear2,percentageOfHypervisorsMigratedOnYear3, reportName, reportDescription,file
+insights.properties=yearOverYearGrowthRatePercentage,percentageOfHypervisorsMigratedOnYear1,percentageOfHypervisorsMigratedOnYear2,percentageOfHypervisorsMigratedOnYear3,reportName,reportDescription
```
Comment (mrizzi) [inline src/main/resources/application.properties:69]: Is this change related to the multi-json?

**Review by mrizzi — COMMENTED** [review]
Reviewed the code (not tests that are WIP).

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 1 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: sure. I dont have a clear criteria to know when it's important and when itt's not. There are several hardcoded texts in the code : identity, internal, user, username, file ...
I will replace the ones accessed in **header** method

Comment (mrizzi) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: Maybe we can start with having the headers we add to be managed with static fields like `ANALYSIS_ID` and `USERNAME`.

Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: ok -> MA_METADATA, UPLOADFORMDATA

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 10 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: +1

File: src/main/resources/application.properties (hunk 11 above)
Comment (jonathanvila) [reply src/main/resources/application.properties:69]: This is to avoid overriding the property in the deployment properties in OCP
Now there is a : insights.properties , with a value without the "file" in the xavier-integration deployment conf in CI.

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 9 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: +1

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 8 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: +1

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 5 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: In fact I will remove this LOG as it was only for testing purposes, with Camel Tracing is enough.

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 4 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: In fact I will remove this LOG as it was only for testing purposes, with Camel Tracing is enough.

File: src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java (hunk 3 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/route/MainRouteBuilder.java]: +1

File: src/main/java/org/jboss/xavier/integrations/migrationanalytics/business/ParamsCalculator.java (hunk 2 above)
Comment (jonathanvila) [reply src/main/java/org/jboss/xavier/integrations/migrationanalytics/business/ParamsCalculator.java]: +1 Good catch

**Review by mrizzi — APPROVED** [review]
Tested in CI with different payloads (single JSON payload, tar.gz with 1 JSON file, tar.gz with 2 JSON files) and everything worked fine.
The code seems good to me after we talked and worked so much about it.
Thanks @jonathanvila for this enhancement :+1:
